### ISSUE TITLE
Incorrect ReturnItem#pre_tax_amount when order contains discounts

### ISSUE THREAD

**Issue by DrPheltRight** [spree/spree#5915]
When a promotion is applied to line item total such as a flat rate discount the generated #pre_tax_amount is including the discount but a post tax discount thus refunding less than it should.

The problem lies in Spree::Calculator::Returns::DefaultRefundAmount#weighted_order_adjustment_amount since we do not proportionately remove tax from the adjustments.

Perhaps this method:

``` ruby
  def weighted_order_adjustment_amount(inventory_unit)
    inventory_unit.order.adjustments.eligible.non_tax.sum(:amount) * percentage_of_order_total(inventory_unit)
  end
```

https://github.com/spree/spree/blob/master/core/app/models/spree/calculator/returns/default_refund_amount.rb#L19

Should be more like this:

``` ruby
  def weighted_order_adjustment_amount(inventory_unit)
    adjustments_after_tax = inventory_unit.order.adjustments.eligible.non_tax.sum(:amount) * percentage_of_order_total(inventory_unit)
    line_item_tax = (inventory_unit.line_item.pre_tax_amount / inventory_unit.line_item.price) * percentage_of_line_item(inventory_unit)
    adjustments_after_tax * line_item_tax
  end
```

I say perhaps since I may be missing something completely obvious!

Am I doing something wrong or is this a bug?

Thanks,
Luke

**Comment by athal7** [comment]
@DrPheltRight i'm having trouble wrapping my head around this one. Do you mind giving an example that contains some numbers? That would help me work through this more easily.

**Comment by DrPheltRight** [comment]
Sure, I have a real example, the numbers aren't pretty but here we go:

```
list_item_1:
  pre_tax: 124.17
  price: 149

list_item_2:
  pre_tax: 45.83
  price: 55

order:
  sub_total: 204
  tax: 34
  promotion_adjustment: -81.60
  total: 122.40
 
tax_rate: 20%
line_item_discount: 40%

calculated_pretax_refund: # this is what is shown on return_authorizations#new
  line_item_1: 64.57
  line_item_2: 23.83

expected_pretax_refund: # this is what I expected to see
  line_item_1: 74.53
  line_item_2: 27.47
```

The reason the pretax refund did not show the right amount is because the discount taken from the pre-tax amount was post tax.

Does that help at all?

**Comment by DrPheltRight** [comment]
My suggested solution above caused a further break in Spree::ReimbursementTaxCalculator which makes me think either I'm missing something rather big or Spree::ReimbursementTaxCalculator was built around this broken behaviour.

**Comment by athal7** [comment]
i'm not sure i follow what is incorrect about this calculation, it does:

124.17 - 81.6*(149/204) = 64.57

which applies the order-level promotion adjustment to the inventory unit's weight of the order

**Comment by DrPheltRight** [comment]
The 81.6 discount is based on the order subtotal that includes taxed line items. However Spree::Calculator::Returns::DefaultRefundAmount is trying to calculate the pretax amount. So we are applying the taxed proportion of discount rather than the untaxed proportion of discount. This means we are creating return item pre tax totals lower than they should be.

**Comment by DrPheltRight** [comment]
You can see the error more easily when you sum the return item pre tax totals and then apply tax:

64.57 + 23.83 = 88.4 * 1.2 = 106.08

The actual order total is 122.40.

**Comment by DrPheltRight** [comment]
Using my suggested solution we instead get the correct order total:

74.53 + 27.47 = 102 * 1.2 = 122.40

**Comment by athal7** [comment]
Ah I see the problem, you have spread your promotion across the line items post-tax, which it ought to be spread across the line items for tax-calculations purposes pre-tax.

The line item calculations should be as follows

```
list_item_1:
  pre_tax: 124.17
  taxable_amount: 124.17 - (81.60 * (124.17 / (124.17 + 45.83))) = 64.56 
  price: (124.17 + 64.56 * 0.2) = 137.08

list_item_2:
  pre_tax: 45.83
  taxable_amount: 124.17 - (81.60 * (124.17 / (124.17 + 45.83))) = 23.83
  price: (45.83 + 23.83 * 0.2) = 50.60

order:
  sub_total: 204
  tax: 34
  promotion_adjustment: -81.60
  total: 106.08

tax_rate: 20%
line_item_discount: 40%
```

**Comment by DrPheltRight** [comment]
That's how spree applied the promotion. Are you saying I've set the
promotion up wrong or is this an issue with spree itself?

**Comment by athal7** [comment]
To my understanding the sales tax was improperly calculated

**Comment by DrPheltRight** [comment]
@JDutil Any advice on how to proceed with this?

**Comment by JDutil** [comment]
Do you know why your sales tax is incorrect?  How do you configure your sales tax?  I'm not sure why your tax would be off, and need to know how to reproduce the problem to  be able to investigate.

**Comment by DrPheltRight** [comment]
I'm not sure the sales tax is incorrect. Our prices include tax by the way.

@athal7 Your price calculations are incorrect. The price of the item is 149 and 55 respectively, not the amounts you calculate.

**Comment by DrPheltRight** [comment]
The pretax price is calculated from the actual price since the price includes the tax (using included_in_price: true for Spree::TaxRate).

**Comment by athal7** [comment]
@DrPheltRight ah that is new information that you are using inclusive tax rather than additional tax. It is entirely possible that the calculations are incorrect for that, as it was developed and tested primarily with additional tax in mind. If you do submit a pull request, please keep in mind that both will need to be supported.

**Comment by JDutil** [comment]
I believe this is fixed by eb4727211869f486c27d96180622e46cf7f8efd9
